---
title: "An introduction to GENESPACE"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(GENESPACE)
```

# 1. Setup

### 1.1 Download raw annotation files

Here we are going to building a synteny-constrained pan-genome annotation with a human, mouse and dog genome. Download these from NCBI as follows:

- Human [peptide fasta](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_translated_cds.faa.gz), [gene gff3](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz)
- Mouse [peptide fasta](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/635/GCA_000001635.9_GRCm39/GCA_000001635.9_GRCm39_translated_cds.faa.gz), [gene gff3](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/635/GCA_000001635.9_GRCm39/GCA_000001635.9_GRCm39_genomic.gff.gz)
- Dog [peptide fasta](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/285/GCF_000002285.5_Dog10K_Boxer_Tasha/GCF_000002285.5_Dog10K_Boxer_Tasha_translated_cds.faa.gz), [gene gff3](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/285/GCF_000002285.5_Dog10K_Boxer_Tasha/GCF_000002285.5_Dog10K_Boxer_Tasha_genomic.gff.gz)

### 1.2 Setup directories

The raw annotation files should be stored somewhere with read/write access that is not purged. For example `~/Desktop/rawAnnotationRepo`. 

The files for each species and genome version need to be moved into their own (sub)directory. So, for the above genomes, make the following directories and move the two annotation files into these. Ensure that the files remain gzip compressed.

Human: `~/Desktop/mammalGenespace/rawRepo/Homo_sapiens/GRCh38.p13/annotation`

Mouse: `~/Desktop/mammalGenespace/rawRepo/Mus_musculus/GRCm39/annotation`

Dog: `~/Desktop/mammalGenespace/rawRepo/Canis_lupus_familiaris/Dog10K_Boxer_Tasha/annotation`

### 1.3 Dependency installation

See the readme for details about dependencies and how to install them. Be sure that you have installed and know the path to MCScanX. Also, in a separate terminal, open an environment that has orthofinder in the path. This is easiest to do as a conda environment, but many other options also work. 



# 2. Initializing a genespace run

## 2.1 Parameterizing the run

After installing, the first step of GENESPACE is to initialize a run. The command `init_genespace` accomplishes this by generating a `gsParam` list that contains all necessary parameters for the genespace run. There are three major sets of arguments for `init_genespace`:

###  Paths to files, programs and directories:

- `wd`: the working directory, where all output will be stored. If this directory does not exist, GENESPACE will make a new directory, assuming the path is valid. 
- `rawGenomeDir`: see above, the location where all raw annotataion files are stored. In this case `~/Desktop/mammalGenespace/rawRepo/`. 
- `path2mcscanx`: the path to the MCScanX executable. This must be valid. If the install is not in the specified path, `init_genespace` returns an error. 
- `path2orthofinder`: the path to the `orthofinder` executable. This is not required to be a valid path. If `init_genespace` cannot find a valid install, it will notify you that orthofinder must be run outside or R. We recommend this anyways. 
- `gffString`: a regular expression or word that identifies the gff file in its directory. If only the two files from above are in the annotation directory, then setting to "gff" will be just fine. If other files are in the directory (e.g. repeat annotations, etc.) you may need more specificity. 
- `pepString`: like `gffString` but for the peptide fasta annotation file. 

### Genome ID specifications

All of these parameters must have the same length (i.e. three genomes, must each have their own genomeID, speciesID, versionID and ploidy)

- `genomeIDs`: unique identifier for each genome. Can be anything you'd like as only as each genomeID is unique, there are no special characters (e.g. *|, etc.) and the first character is a letter. This is used as the identifier in all plots and output, in lieu of the genome version and species ID. 
- `speciesIDs`: this is the directory within `rawGenomeDir` that contains the annotations for each genomeID. This does not necessarily need to be unique (i.e. you want to compare two assemblies from one species).
- `versionIDs`: this is the directory within`rawGenomeDir/speciesID` that contains the annotation subdirectory, which contains the annotation files. 
- `ploidy`: the ploidy of the genome assembly, not necessarily the organism. In this case, all three species are diploid, but their genome assemblies are haploid. 

### Other global parameters

- `nCores`: the number of parallel processes to run. If left out, GENESPACE will choose half of the available cores. Ensure that you are working with the right number of cores when on a cluster, as, often the number of detected cores does not match the number available to the user.
- `minPepLen`: the minimum peptide length to be kept in the annotation files. Defaults to 20. 

```
gpar <- init_genespace(
  genomeIDs = c("human", "mouse", "dog"),
  speciesIDs = c("Homo_sapiens", "Mus_musculus", "Canis_lupus_familiaris"),
  versionIDs = c("GRCh38.p13", "GRCm39", "Dog10K_Boxer_Tasha"),
  ploidy = c(1, 1, 1),
  wd = "~/Desktop/exampleGenespace",
  nCores = 4, 
  gffString = "gff", 
  pepString = "translated_cds",
  path2orthofinder = "NULL",
  path2mcscanx = "~/Documents/comparative_genomics/programs/MCScanX",
  rawGenomeDir = "~/Desktop/mammalGenespace/rawRepo")

# set working directory to ~/Desktop/exampleGenespace 
# 
# found raw gff files:
#   ~/Desktop/mammalGenespace/rawRepo/Homo_sapiens/GRCh38.p13/annotation/GCF_000001405.39_GRCh38.p13_genomic.gff.gz
# ~/Desktop/mammalGenespace/rawRepo/Mus_musculus/GRCm39/annotation/GCF_000001635.27_GRCm39_genomic.gff.gz
# ~/Desktop/mammalGenespace/rawRepo/Canis_lupus_familiaris/Dog10K_Boxer_Tasha/annotation/GCF_000002285.5_Dog10K_Boxer_Tasha_genomic.gff.gz
# 
# found raw peptide files:
#   ~/Desktop/mammalGenespace/rawRepo/Homo_sapiens/GRCh38.p13/annotation/GCF_000001405.39_GRCh38.p13_translated_cds.faa.gz
# ~/Desktop/mammalGenespace/rawRepo/Mus_musculus/GRCm39/annotation/GCF_000001635.27_GRCm39_translated_cds.faa.gz
# ~/Desktop/mammalGenespace/rawRepo/Canis_lupus_familiaris/Dog10K_Boxer_Tasha/annotation/GCF_000002285.5_Dog10K_Boxer_Tasha_translated_cds.faa.gz 
# 
# GENESPACE run initialized
# 
# ######***NOTE***########
# You will need to conduct an orthofinder run prior to using genespace
# Follow instructions printed from run_orthofinder
# Warning message:
#   In check_orthofinderInstall(path2orthofinder) :
#   Cannot call orthofinder from path specified to GENESPACE
# Run orthofinder outside of R with commands supplied
```


## 2.2 Formatting annotations

GENESPACE acts on parsed gff3-like text files and peptide fasta annotations that have headers that exactly match the `id` column in the gff3-like text file. While it is possible to make these manually, we offer a set of convenience functions. There are two pre-optimized methods: `parse_ncbi` and `parse_phytozome` which format NCBI and phytozome annotations with no user-specified parameters. However, it may be that the annotations of interest come from a different source. As long as the annotations are peptide fastas and gff3-formatted text files, it should be possible to parse each using rules in `parse_annotations`. 

In the case where some of your genomes are formatted in different ways, each parsing function has a `genomeIDs` parameter that allow you to apply that parsing method only for those genomes. 

#### Annotation parsing parameters:

`parse_annotations` processes the gff3 `attributes` column and the fasta header string so that they match. 

- `gffEntryType`: the value in the `type` (3rd) gff column to subset to. Usually "gene" or "mRNA". 
- `gffIdColum`: the ";" separated field in the attribute column to use as the ID
- `headerSep`: the delimiter in the fasta header to split different attributes
- `headerEntryIndex`: which value in the separated header to use
- `headerStripText`: any text that needs to be removed from the parsed sequence so that it will match with the gff. 

It can sometimes be hard to decide what rules work for your annotation of interest. Setting `troubleshoot = TRUE` will print parsed and un-parsed annotations too let the user see if it is working. 

```
parse_ncbi(gsParam = gpar, overwrite = T)

# Parsing annotations: human
# Reading gff ... found 19532 protein coding genes
# Reading peptide fasta ... found 124038 / 19532 total and unique entries
# Merging fa and gff found 19532 matching entires

# Parsing annotations: mouse
# Reading gff ... found 22186 protein coding genes
# Reading peptide fasta ... found 93165 / 22186 total and unique entries
# Merging fa and gff found 22186 matching entires

# Parsing annotations: dog
# Reading gff ... found 20087 protein coding genes
# Reading peptide fasta ... found 62437 / 20087 total and unique entries
# Merging fa and gff found 20087 matching entires
```


### 2.3 Set the synteny parameters

GENESPACE is predicated on synteny constrained blast(like) hits, which are found using `synteny`. In most cases, the default synteny parameters will do just fine and can be added to the parameter list with `gpar <- set_syntenyParams(gsParam = gpar)`. However, some genomes need special care. Some considerations:

- Massive tandem arrays: One step of synteny constraining is masking regions around self blast hits so that tandem arrays do not appear as their own syntenic regions (MCScanX is likely to find many small syntenic blocks within large arrays). GENESPACE controls for this with the `selfRegionMask` parameter, which defines how far, in gene rank order position, away from self hits a new syntenic region can be formed.
- Loose synteny (e.g. tiny blocks some distance away from the main syntenic hits): except in very simple systems, synteny is never just a one-to-one blast hit 'diagonal'. Often, tandem arrays, small translocations or inversion disrupt synteny. GENESPACE controls for this by pulling blast hits with a 'buffer' of the primary collinear anchors derived from MCScanX, which is specified as the `synBuff`. 
- Lots of very small blocks. GENESPACE pases the `blkSize` (n. genes in a block) and `nGaps` (number of non-adjacencies permitted within a block) to MCScanX. Higher values of `blkSize` and lower vaalaues of `nGaps` increases block calling stringency. The opposite allows inclusion of smaller more discontiguous blocks.

For more detail, see the description of the synteny pruning below or `?set_syntenyParams`.

```
gpar <- set_syntenyParams(gpar)
print(gpar$params$synteny)

# genome1 genome2           u ploidy1 ploidy2 ofID1 ofID2 nGenes1 nGenes2
# 1:   mouse   mouse mouse mouse       1       1     2     2   22186   22186
# 2:     dog   mouse   dog mouse       1       1     0     2   20087   22186
# 3:   mouse     dog   mouse dog       1       1     2     0   22186   20087
# 4:   human   mouse human mouse       1       1     1     2   19532   22186
# 5:   mouse   human mouse human       1       1     2     1   22186   19532
# 6:     dog     dog     dog dog       1       1     0     0   20087   20087
# 7:     dog   human   dog human       1       1     0     1   20087   19532
# 8:   human     dog   human dog       1       1     1     0   19532   20087
# 9:   human   human human human       1       1     1     1   19532   19532
# 
# nhits1 nhits2 query target runBlast mirrorBlast nSecondHits1 nSecondHits2
# 1:      1      1 mouse  mouse     TRUE       FALSE            0            0
# 2:      1      1 mouse    dog    FALSE        TRUE            0            0
# 3:      1      1 mouse    dog     TRUE       FALSE            0            0
# 4:      1      1 mouse  human    FALSE        TRUE            0            0
# 5:      1      1 mouse  human     TRUE       FALSE            0            0
# 6:      1      1   dog    dog     TRUE       FALSE            0            0
# 7:      1      1   dog  human     TRUE       FALSE            0            0
# 8:      1      1   dog  human    FALSE        TRUE            0            0
# 9:      1      1 human  human     TRUE       FALSE            0            0
# 
# onlyOgAnchors onlyOgAnchorsSecond blkSize blkSizeSecond nGaps nGapsSecond
# 1:          TRUE                TRUE       5             5     5           5
# 2:          TRUE                TRUE       5             5     5           5
# 3:          TRUE                TRUE       5             5     5           5
# 4:          TRUE                TRUE       5             5     5           5
# 5:          TRUE                TRUE       5             5     5           5
# 6:          TRUE                TRUE       5             5     5           5
# 7:          TRUE                TRUE       5             5     5           5
# 8:          TRUE                TRUE       5             5     5           5
# 9:          TRUE                TRUE       5             5     5           5
# 
# synBuff synBuffSecond selfRegionMask nSecondHits
# 1:     100           100            200           0
# 2:     100           100            200           0
# 3:     100           100            200           0
# 4:     100           100            200           0
# 5:     100           100            200           0
# 6:     100           100            200           0
# 7:     100           100            200           0
# 8:     100           100            200           0
# 9:     100           100            200           0
# 
# dropInterleavesSmallerThan minRbhScore
# 1:                          2          50
# 2:                          2          50
# 3:                          2          50
# 4:                          2          50
# 5:                          2          50
# 6:                          2          50
# 7:                          2          50
# 8:                          2          50
# 9:                          2          50
```

### 2.4 Run orthofinder

If orthofinder is in the path available to R, the command `run_orthofinder` will actually call orthofinder from R. However, we do not suggest this, since dependencies are often in conflict between R and orthofinder. Here, we are assuming that R is run from an environment without orthofinder in the path; therefore, GENESPACE will output the command that needs to be run, then the user should open a terminal with orthofinder in the path (see README for details on install instructions) and run the following command:

```
gpar <- run_orthofinder(gsParam = gpar, overwrite = F)

# Cleaning out orthofinder directory and prepping run
# Could not find valid orthofinder executable in the path
# Run the following command outside of R (assuming orthofinder is in the path):
# ################
# orthofinder -f /Users/jlovell/Desktop/exampleGenespace/peptide -t 4 -a 1 -X -o ~/Desktop/exampleGenespace/orthofinder
# ################
```

In this case, you would open your orthofinder conda env `source activate orthofinder`, the run `orthofinder -f /Users/jlovell/Desktop/exampleGenespace/peptide -t 4 -a 1 -X -o ~/Desktop/exampleGenespace/orthofinder` in that environment. This produces the raw orthofinder output and blast results on which GENESPACE depends. 



The synteny search runs a three step procedure:

First, 'primary' and (optionally) 'secondary' region anchors are formed. The primary regions are the self hits in intragenomic searches, and the ploidy-informed top N constrained intergenomic hits. For example, when comparing a diploid genome (gen1) to a haploid genome (gen2), the primary search would only see the 




