#' @title See if more mappings can be added
#'
#' @description
#' \code{add2_blocks} Adds mappings to blocks that were not included previously
#'
#' @param blast.results parsed blast results, generated by runParse_orthofinder
#' @param map the map object
#' @param blk the block object
#' @param verbose logical, should updates be printed?
#' @param ... Not currently in use
#' @details Needs to be run prior to the pipeline. Makes some objects that are required.
#' @return Nothing.
#'
#' @examples
#' \dontrun{
#' none yet
#' }
#' @import data.table
#' @export
add2_blocks = function(blast.results, map, blk,
                       gff.dir, verbose = T){
  if(verbose)
    cat("Parsing gff files\n")
  gff.files <- list.files(gff.dir,
                          full.names = T)
  names(gff.files) <- gsub(".gff3$", "",
                           basename(gff.files))
  parse_gff <- function(gff){
    g <- suppressWarnings(
      data.table::fread(gff,
                        showProgress = F,
                        verbose = F))
    g <- g[g$V3 == "gene", c(9, 1, 4, 5, 7)]
    g$V9 <- sapply(g$V9, function(x) gsub("Name=", "",
                                          strsplit(x, ";")[[1]][2]))
    data.table::setnames(g, c("id", "chr", "start", "end", "strand"))
    return(g)
  }
  gff.list <- sapply(names(gff.files), simplify = F, USE.NAMES = T, function(i){
    tmp <- parse_gff(gff.files[[i]])
    tmp$genome <- i
    tmp$order <- frank(tmp[,c("chr", "start")],
                       ties.method = "dense")
    return(tmp)
  })
  if(verbose)
    cat("Adding blast hits into map file ... initial n hits ==", nrow(map),"\n")
  ret = rbindlist(lapply(1:nrow(blk), function(i){
    x = blk[i,]
    y1 = gff.list[[x$genome1]]
    y2 = gff.list[[x$genome2]]
    y1 = y1[with(y1, chr == x$chr1 & end >= x$start1 & start <= x$end1),]
    y2 = y2[with(y2, chr == x$chr2 & end >= x$start2 & start <= x$end2),]
    out = blast.results[with(blast.results, id1 %in% y1$id & id2 %in% y2$id),]
    out$block.id = x$block.id
    out$mapping = paste(x$genome1, x$genome2)
    return(out[,colnames(map),with =F])
  }))
  if(verbose)
    cat("Done! Returning map object with", nrow(ret),"hits\n")
  return(ret)
}
