#' @title Make a circos plot
#'
#' @description
#' \code{plot_linkedCircos} Make a circos plot with links connecting
#' syntenic / collinear blocks
#'
#' @param blk The block data.frame or data.table
#' @param genome_id1 Identifier of the first genome to plot
#' @param genome_id2 Identifier of the second genome to plot
#' @param chrs1 Chromosome IDs of the genome_id1 to plot. Order is retained.
#' @param chrs2 Chromosome IDs of the genome_id2 to plot. Order is retained.
#' @param chr_order Numeric vector to reorder above set of chromsomes.
#' @param chrs1_abbrev The plotted names of chrs1
#' @param chrs2_abbrev The plotted names of chrs2
#' @param cols Character vector of colors, must match chr_order length
#' @param gap.deg Vector of gap degrees between chromosomes. Must match chr_order length
#' @param assembly.dir The directory containing the assembly fasta indices. If not
#' generated by `convert_genomes`, the indices must have a name of
#' assembly.dir/genomeID.fa.fai.
#' @param genome_id1.col The color of a track just inside of the labels, indicating
#' that the chromosome comes from genomeID1.
#' @param genome_id2.col The color of a track just inside of the labels, indicating
#' that the chromosome comes from genomeID2.
#' @param track.height How wide should the track with genome colors be?
#' @param adjust.alpha Logical, should transparency be adjusted based on block size?
#' @param link.alpha Numeric 0-1, specify opacity of links.
#' @param ... Not currently in use
#' @details More here
#' @return Nothing.
#'
#' @examples
#' \dontrun{
#' plot_linkedCircos(blk = mcs.results$block,
#' assembly.dir = assembly.dir,
#' genome_id1 = "PhalliiHAL",
#' genome_id2 = "Pvirgatum",
#' chrs1 = paste0("Chr0",1:9),
#' chrs2 = c(paste0("Chr0",1:9,"K"),paste0("Chr0",1:9,"N")),
#' chrs1_abbrev = paste("Ph",1:9),
#' chrs2_abbrev = c(paste0("Pv ",1:9,"K"),paste0("Pv ",1:9,"N")),
#' chr_order = c(1,10,19,2,11,20,3,12,21,4,13,22,5,14,23,6,15,24,7,16,25,8,17,26,9,18,27),
#' cols = rep(c("indianred4","red3","darkorange2",
#'              "gold3","chartreuse4","darkturquoise",
#'               "dodgerblue2","purple3","magenta3"),
#'            each = 3),
#' gap.deg = rep(c(.5,.5, 4), 9))
#' }
#' @import data.table
#' @importFrom circlize circos.par circos.genomicInitialize circos.track circos.genomicLink circos.clear
#' @export
plot_linkedCircos <- function(blk,
                              genomes,
                              chr.order.list,
                              genome_id1,
                              genome_id2,
                              chrs1,
                              chrs2,
                              chr_order,
                              chrs1_abbrev,
                              chrs2_abbrev,
                              cols,
                              gap.deg,
                              assembly.dir,
                              genome_id1.col = "grey80",
                              genome_id2.col = "grey50",
                              track.height = 0.05,
                              adjust.alpha = F,
                              link.alpha = .6,
                              ...){

  blk <- data.table(blk)

  add.alpha <- function(col, alpha = 1){
    if (missing(col))
      stop("Please provide a vector of colours.")

    apply(sapply(col, col2rgb)/255, 2,
          function(x)
            rgb(x[1],
                x[2],
                x[3],
                alpha = alpha))
  }

  fai1 <- read.delim(file.path(assembly.dir,
                               paste0(genome_id1,
                                      ".fa.fai")),
                     header = F,
                     stringsAsFactors = F,
                     col.names = c("chr", "length", "v1", "v2", "v3"))[,1:2]

  fai1 <- fai1[fai1$chr %in% chrs1,]
  rownames(fai1) <- fai1$chr
  fai1 <- fai1[chrs1,]
  fai1$sector_names <- chrs1_abbrev
  fai1$genome <- genome_id1

  fai2 <- read.delim(file.path(assembly.dir,
                               paste0(genome_id2, ".fa.fai")),
                     header = F,
                     stringsAsFactors = F,
                     col.names = c("chr", "length", "v1", "v2", "v3"))[,1:2]

  fai2 <- fai2[fai2$chr %in% chrs2,]
  rownames(fai2) <- fai2$chr
  fai2 <- fai2[chrs2,]
  fai2$sector_names <- chrs2_abbrev
  fai2$genome <- genome_id2
  fais <- rbind(fai1, fai2)
  fais <- fais[chr_order,]
  fais$color <- cols


  setkey(blk,
         chr1, chr2,
         start1, start2)

  blk <- data.frame(blk,
                    stringsAsFactors = F)
  b0 <- blk[with(blk, (genome1 == genome_id1 |
                         genome2 == genome_id1) &
                   (genome1 == genome_id2 |
                      genome2 == genome_id2)),]
  b1 <- b0[with(b0, genome1 == genome_id1),]
  b2 <- b0[with(b0, genome1 == genome_id2),]

  good_colnames <- c("genome1", "genome2",
                     "chr1", "chr2",
                     "start1", "start2",
                     "end1", "end2",
                     "orient")
  colnames_2switch <- c("genome2", "genome1",
                        "chr2", "chr1",
                        "start2", "start1",
                        "end2", "end1",
                        "orient")
  if (nrow(b2) == 0) {
    bed <- b1[, good_colnames]
  }else{
    if (nrow(b1) == 0) {
      bed <- b2[, colnames_2switch]
      colnames(bed) <- good_colnames
    }else{
      b1t <- b1[, good_colnames]
      b2t <- b2[, colnames_2switch]
      colnames(b2t) <- good_colnames
      bed <- rbind(b1t, b2t)
    }
  }
  bed <- bed[bed$chr1 %in% fai1$chr &
               bed$chr2 %in% fai2$chr,]


  bed$s1 <- with(bed,
                 ifelse(orient == "+", start1, end1))
  bed$e1 <- with(bed,
                 ifelse(orient == "+", end1, start1))
  bed$start1 <- bed$s1
  bed$end1 <- bed$e1
  bed$s1 <- NULL
  bed$e1 <- NULL


  if (adjust.alpha) {
    alpha.scale <- function(x){
      ifelse(x > 5e6,.6,
             ifelse(x > 1e6,.4,
                    ifelse(x > 5e5, .3,
                           ifelse(x > 1e5,.2,.1))))
    }
  }else{
    alpha.scale <- function(x){
      return(link.alpha)
    }
  }

  bed1 <- with(bed,
               data.frame(chr = chr1,
                          start = start1,
                          end = end1,
                          value = alpha.scale(abs(start1 - end1)),
                          color = NA,
                          stringsAsFactors = F))
  bed2 = with(bed,
              data.frame(chr = chr2,
                         start = start2,
                         end = end2,
                         value = alpha.scale(abs(start2 - end2)),
                         color = NA,
                         stringsAsFactors = F))
  for (i in 1:length(chrs1)) {
    bed1$color[bed1$chr == chrs1[i]] <- fais$color[fais$chr == chrs1[i] &
                                                     fais$genome == genome_id1]
    bed1$chr[bed1$chr == chrs1[i]] <- chrs1_abbrev[i]
  }

  for (i in 1:length(chrs2)) {
    bed2$color[bed2$chr == chrs2[i]] <- fais$color[fais$chr == chrs2[i] &
                                                     fais$genome == genome_id2]
    bed2$chr[bed2$chr == chrs2[i]] <- chrs2_abbrev[i]
  }

  bed1$link.col <- sapply(1:nrow(bed1), function(x)
    add.alpha(bed1$color[x],
              bed1$value[x]))
  bed2$link.col <- sapply(1:nrow(bed2), function(x)
    add.alpha(bed2$color[x],
              bed2$value[x]))

  linkcols <- bed1$link.col

  fais$color <- ifelse(fais$genome == genome_id1,
                       genome_id1.col, genome_id2.col)

  init <- data.frame(name = fais$sector_names,
                     start = 0,
                     end = fais$length,
                     stringsAsFactors = F)
  circos.par(start.degree = 90,
             clock.wise = TRUE,
             "gap.degree" = gap.deg)
  circos.genomicInitialize(init)

  circos.track(ylim = c(0, 1),
               bg.col = fais$color,
               bg.border = NA,
               track.height = track.height)
  circos.genomicLink(bed1,
                     bed2,
                     col = linkcols,
                     border = NA)
  circos.clear()
}
